### ISSUE TITLE
build: linker error on SUSE Linux with Sun C 5.15

### ISSUE THREAD

**Issue by KavithaTipturMadhu** [pmodels/mpich#3358]
Linker error building mpich master mpich-master on "SUSE Linux Enterprise Server 12.3 (x86_64)" with Sun C 5.15 (Oracle Developer Studio 12.6).

Reported by Siegmar Gross <siegmar.gross@informatik.hs-fulda.de> when building MPICH master.

```

loki mpich-master_2018-10-05_Linux.x86_64.64_cc 128 head -7 config.log | tail -1
   $ ../mpich/configure --prefix=/usr/local/mpich-master_64_cc 
--libdir=/usr/local/mpich-master_64_cc/lib64 
--includedir=/usr/local/mpich-master_64_cc/include64 CC=cc CXX=CC F77=f77 FC=f95 
CFLAGS=-m64 CXXFLAGS=-m64 FFLAGS=-m64 FCFLAGS=-m64 LDFLAGS=-m64 -L/usr/lib64 
-R/usr/lib64 --enable-fortran=no --enable-cxx --enable-romio --enable-debuginfo 
--enable-smpcoll --enable-threads=multiple --with-thread-package=posix 
--enable-shared

loki mpich-master_2018-10-05_Linux.x86_64.64_cc 128 tail -40 
log.make.Linux.x86_64.64_cc
"/export2/src/mpich-master/mpich/src/openpa/src/primitives/opa_gcc_intel_32_64_ops.h", 
line 162: warning: parameter in inline asm statement unused: %3
"../mpich/src/include/mpir_handlemem.h", line 61: warning: null dimension: indirect
"../mpich/src/include/mpir_handlemem.h", line 122: warning: null dimension: indirect
   CC       src/env/mpivars.o
   CCLD     lib/libtvmpich.la
   GEN      lib/libmpi.la
"./src/include/mpichconf.h", line 1718: Warning (Anachronism): Attempt to 
redefine __restrict__ without using #undef.
1 Warning(s) detected.
   CXXLD    lib/libmpicxx.la
   CCLD     src/env/mpichversion
   CCLD     src/env/mpivars
CC: -library=Crun cannot be used with -std=c++03. To use this library you need 
to switch to -std=sun03
CC: -library=Cstd cannot be used with -std=c++03. To use this library you need 
to switch to -std=sun03
Makefile:10783: recipe for target 'lib/libmpicxx.la' failed
make[2]: *** [lib/libmpicxx.la] Error 1
make[2]: *** Waiting for unfinished jobs....
src/env/mpichversion.o: In function `hwloc_distrib':
mpichversion.c:(.text+0xb0): undefined reference to `hwloc_bitmap_weight'
mpichversion.c:(.text+0x13d): undefined reference to `hwloc_bitmap_weight'
mpichversion.c:(.text+0x1e4): undefined reference to `hwloc_bitmap_dup'
mpichversion.c:(.text+0x219): undefined reference to `hwloc_bitmap_or'
src/env/mpichversion.o: In function `MPIR_Datatype_free_contents':
mpichversion.c:(.text+0x30d): undefined reference to `MPIR_Datatype_builtin'
mpichversion.c:(.text+0x32b): undefined reference to `MPIR_Datatype_direct'
mpichversion.c:(.text+0x339): undefined reference to `MPIR_Datatype_mem'
mpichversion.c:(.text+0x37f): undefined reference to `MPIR_Process'
mpichversion.c:(.text+0x3f9): undefined reference to `MPIR_Assert_fail'
src/env/mpichversion.o: In function `MPIR_Datatype_free':
mpichversion.c:(.text+0x43e): undefined reference to `MPIR_Dataloop_free'
mpichversion.c:(.text+0x445): undefined reference to `MPIR_Datatype_mem'
postopt: error: ld failed to link the binary
cc: postopt failed for src/env/.libs/mpichversion
Makefile:20510: recipe for target 'src/env/mpichversion' failed
make[2]: *** [src/env/mpichversion] Error 2
make[2]: Leaving directory 
'/export2/src/mpich-master/mpich-master_2018-10-05_Linux.x86_64.64_cc'
Makefile:40490: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory 
'/export2/src/mpich-master/mpich-master_2018-10-05_Linux.x86_64.64_cc'
Makefile:10578: recipe for target 'all' failed
make: *** [all] Error 2
loki mpich-master_2018-10-05_Linux.x86_64.64_cc 129`
```

Update:
Another opensuse related issue was reported by the same person that reported the previous issue.
```loki mpich-master-v3.3b3-357-gdcd523e-Linux.x86_64.64_cc 138 tail -30 log.make.Linux.x86_64.64_cc | grep -v NOTICE
"../mpich-master-v3.3b3-357-gdcd523e/src/include/mpir_handlemem.h", line 122: warning: null dimension: indirect
 CC       src/pmi/simple/lib_libmpi_la-simple_pmiutil.lo
 CC       src/pmi/simple/lib_libmpi_la-simple_pmi.lo
 CC       src/mpi/debugger/src_mpi_debugger_libdbginitdummy_la-dbginit.lo
"/export2/src/mpich-3.3/mpich-master-v3.3b3-357-gdcd523e/src/openpa/src/primitives/opa_gcc_intel_32_64_ops.h", line 98: warning: parameter in inline asm statement unused: %2
"/export2/src/mpich-3.3/mpich-master-v3.3b3-357-gdcd523e/src/openpa/src/primitives/opa_gcc_intel_32_64_ops.h", line 105: warning: parameter in inline asm statement unused: %1
"/export2/src/mpich-3.3/mpich-master-v3.3b3-357-gdcd523e/src/openpa/src/primitives/opa_gcc_intel_32_64_ops.h", line 112: warning: parameter in inline asm statement unused: %1
"/export2/src/mpich-3.3/mpich-master-v3.3b3-357-gdcd523e/src/openpa/src/primitives/opa_gcc_intel_32_64_ops.h", line 121: warning: parameter in inline asm statement unused: %2
"/export2/src/mpich-3.3/mpich-master-v3.3b3-357-gdcd523e/src/openpa/src/primitives/opa_gcc_intel_32_64_ops.h", line 128: warning: parameter in inline asm statement unused: %3
"/export2/src/mpich-3.3/mpich-master-v3.3b3-357-gdcd523e/src/openpa/src/primitives/opa_gcc_intel_32_64_ops.h", line 140: warning: parameter in inline asm statement unused: %1
"/export2/src/mpich-3.3/mpich-master-v3.3b3-357-gdcd523e/src/openpa/src/primitives/opa_gcc_intel_32_64_ops.h", line 148: warning: parameter in inline asm statement unused: %1
"/export2/src/mpich-3.3/mpich-master-v3.3b3-357-gdcd523e/src/openpa/src/primitives/opa_gcc_intel_32_64_ops.h", line 155: warning: parameter in inline asm statement unused: %3
"/export2/src/mpich-3.3/mpich-master-v3.3b3-357-gdcd523e/src/openpa/src/primitives/opa_gcc_intel_32_64_ops.h", line 162: warning: parameter in inline asm statement unused: %3
"../mpich-master-v3.3b3-357-gdcd523e/src/include/mpir_handlemem.h", line 61: warning: null dimension: indirect
"../mpich-master-v3.3b3-357-gdcd523e/src/include/mpir_handlemem.h", line 122: warning: null dimension: indirect
 CCLD     src/mpi/debugger/libdbginitdummy.la
 GEN      lib/libmpi.la
f90: Warning: Option -shared passed to ld, if ld is invoked, ignored otherwise
/usr/bin/ld: /opt/sun/developerstudio12.6/lib/compilers/amd64/crt1x.o: relocation R_X86_64_32 against hidden symbol `__dso_handle' can not be used when making a shared object
/usr/bin/ld: final link failed: nonrepresentable section on output
postopt: error: ld failed to link the binary
Makefile:17007: recipe for target 'lib/libmpi.la' failed
make[2]: *** [lib/libmpi.la] Error 2
make[2]: Leaving directory '/export2/src/mpich-3.3/mpich-master-v3.3b3-357-gdcd523e-Linux.x86_64.64_cc'
Makefile:40517: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/export2/src/mpich-3.3/mpich-master-v3.3b3-357-gdcd523e-Linux.x86_64.64_cc'
Makefile:10583: recipe for target 'all' failed
make: *** [all] Error 2
loki mpich-master-v3.3b3-357-gdcd523e-Linux.x86_64.64_cc 138


loki mpich-master-v3.3b3-357-gdcd523e-Linux.x86_64.64_cc 141 head -7 config.log | tail -1
 $ ../mpich-master-v3.3b3-357-gdcd523e/configure --prefix=/usr/local/mpich-3.3_64_cc --libdir=/usr/local/mpich-3.3_64_cc/lib64 --includedir=/usr/local/mpich-3.3_64_cc/include64 CC=cc CXX=CC F77=f77 FC=f95 CFLAGS=-m64 CXXFLAGS=-m64 FFLAGS=-m64 FCFLAGS=-m64 LDFLAGS=-m64 -L/usr/lib64 -R/usr/lib64 --enable-fortran=yes --enable-cxx --enable-romio --enable-debuginfo --enable-smpcoll --enable-threads=multiple --with-thread-package=posix --enable-shared
loki mpich-master-v3.3b3-357-gdcd523e-Linux.x86_64.64_cc 142
```

**Comment by gcongiu** [comment]
@KavithaTipturMadhu I have also seen the same issue (# 2) but did not investigated further. If I recall correctly @raffenet mentioned he came across this same problem in the past.

**Comment by KavithaTipturMadhu** [comment]
> @KavithaTipturMadhu I have also seen the same issue (# 2) but did not investigated further. If I recall correctly @raffenet mentioned he came across this same problem in the past (?)

I will try to reproduce the first issue, and link second one to the other PR, since that might be specific to Oracle Developer Studio. Thanks @gcongiu for the heads up.

**Comment by raffenet** [comment]
Should now be fixed with #4122
